src/facade/arg_parser.h

```diff
@@ -0,0 +1,176 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <absl/strings/match.h>
+
+#include <string_view>
+
+#include "facade/error.h"
+#include "facade/facade_types.h"
+
+namespace facade {
+
+struct ArgumentParser {
+  struct NextProxy;
+  friend struct NextProxy;
+  struct CheckProxy;
+  friend struct CheckProxy;
+
+  enum ErrorType {
+    OUT_OF_BOUNDS,
+    SHORT_OPT_TAIL,
+    INVALID_INT,
+    INVALID_CASES,
+  };
+
+  struct NextProxy {
+    operator std::string_view() {
+      return ToSV(parser_->args_[idx_]);
+    }
+
+    operator std::string() {
+      return std::string{operator std::string_view()};
+    }
+
+    template <typename T> operator T() {
+      T out;
+      if (absl::SimpleAtoi(operator std::string_view(), &out))
+        return out;
+      parser_->Report(INVALID_INT, idx_);
+      return T{0};
+    }
+
+    template <typename T> T Cases(std::initializer_list<std::pair<std::string_view, T>> values) {
+      std::string_view arg = operator std::string_view();
+      for (const auto& [tag, value] : values) {
+        if (absl::EqualsIgnoreCase(tag, arg))
+          return value;
+      }
+      parser_->Report(INVALID_CASES, idx_);
+      return T{};
+    }
+
+   private:
+    friend struct ArgumentParser;
+
+    NextProxy(ArgumentParser* parser, size_t idx) : parser_{parser}, idx_{idx} {
+    }
+
+    ArgumentParser* parser_;
+    size_t idx_;
+  };
+
+  struct CheckProxy {
+    operator bool() {
```

Comment:
```suggestion
    explicit operator bool() {
```

Reply:
The idea if to use it in if statements:

 ```c++
if (parser.Check("ON").ExpectTail(1) {
  // handle ON clause here
}
```

Reply:
Many things :)
https://www.artima.com/articles/the-safe-bool-idiom
But if made explicit that's OK

Reply:
Examples:
```cpp
  CheckProxy test;
  test << 1;
  int i=test;
  if (test < test) { .. .}
```

Reply:
it still works: https://godbolt.org/z/xPEs3r1oT by the magic of 'contextual conversion to bool'

Reply:
Are those things you would do with a chaining proxy object? 🙂 Is it something that would be unnoticed? Does it cause invalid access to the proxy object? 

I see how hard-to-read code has an impact on development and is prone to bugs. I don't see how writing 
```c++
bool (parser.Check().ExpectTail())
``` 

makes our code much safer. As a trade-off, I can add an `Ok()` function